src/sagemaker/jumpstart/curated_hub/curated_hub.py

```diff
@@ -113,3 +146,159 @@ def delete_model(self, model_name: str, model_version: str = "*") -> None:
+                return True
+        return False
+
+    def sync(self, model_list: List[Dict[str, str]]):
+        """Syncs a list of JumpStart model ids and versions with a CuratedHub
+
+        Args:
+            model_list (List[Dict[str, str]]): List of `{ model_id: str, version: Optional[str] }`
+                objects that should be synced into the Hub.
+        """
+        if self._is_invalid_model_list_input(model_list):
+            raise ValueError(
+                "Model list should be a list of objects with values 'model_id',",
+                "and optional 'version'.",
+            )
+
+        # Fetch required information
+        # self._get_studio_manifest_map()
+        hub_models = self.list_models()
+
+        # Retrieve latest version of unspecified JumpStart model versions
+        model_version_list = []
+        for model in model_list:
+            # TODO: Uncomment and implement
+            # if not model["version"] or model["version"] == "*":
+            #     model["version"] = self._find_latest_version(model_name=model.model_id)
+            model_version_list.append(model)
+
+        # Find synced JumpStart model versions in the Hub
+        js_models_in_hub = []
+        for model in hub_models:
+            # TODO: extract both in one pass
+            jumpstart_model_id = next(
+                (tag for tag in model.search_keywords if tag.startswith("@jumpstart-model-id")),
+                None,
+            )
+            jumpstart_model_version = next(
+                (
+                    tag
+                    for tag in model.search_keywords
+                    if tag.startswith("@jumpstart-model-version")
+                ),
+                None,
+            )
+
+            if jumpstart_model_id and jumpstart_model_version:
+                js_models_in_hub.append(model)
+
+        # Match inputted list of model versions with synced JumpStart model versions in the Hub
+        models_to_sync = []
+        for model in model_version_list:
+            model_id, version = model
+            matched_model = next((model for model in js_models_in_hub if model.name == model_id))
+
+            # Model does not exist in Hub, sync
+            if not matched_model:
+                models_to_sync.append(model)
+
+            if matched_model:
+                # 1. Model version exists in Hub, pass
+                if matched_model.version == version:
+                    pass
+
+                # 2. Invalid model version exists in Hub, pass
+                # This will only happen if something goes wrong in our metadata
+                if matched_model.version > version:
+                    pass
+
+                # 3. Old model version exists in Hub, update
+                if matched_model.version < version:
+                    # Check minSDKVersion against current SDK version, emit log
+                    models_to_sync.append(model)
+
```

Comment:
If we accept the use case of someone wanting to downgrade a model version, I think it would high level simplify down to `models_to_sync = set(hub_models) - set(model_list)` -> feed the results into the executor.

Reply:
+1, it seems like a good candidate for set operations Ben. Right now, you're performing list search in a loop, which is a recipe for trouble.